tests/test_resource.h

```diff
@@ -0,0 +1,114 @@
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef TEST_RESOURCE
+#define TEST_RESOURCE
+
+#include "core/io/resource.h"
+#include "core/io/resource_loader.h"
+#include "core/io/resource_saver.h"
+#include "core/os/os.h"
+
+#include "thirdparty/doctest/doctest.h"
+
+namespace TestResource {
+
+TEST_CASE("[Resource] Duplication") {
+	Ref<Resource> resource = memnew(Resource);
+	resource->set_name("Hello world");
+	Ref<Resource> child_resource = memnew(Resource);
+	child_resource->set_name("I'm a child resource");
+	resource->set_meta("other_resource", child_resource);
+
+	Ref<Resource> resource_dupe = resource->duplicate();
+	const Ref<Resource> resource_dupe_reference = resource_dupe;
+	resource_dupe->set_name("Changed name");
+	child_resource->set_name("My name was changed too");
+
+	CHECK_MESSAGE(
+			resource_dupe->get_name() == "Changed name",
+			"Duplicated resource should have the new name.");
+	CHECK_MESSAGE(
+			resource_dupe_reference->get_name() == "Changed name",
+			"Reference to the duplicated resource should have the new name.");
+	CHECK_MESSAGE(
+			resource->get_name() == "Hello world",
+			"Original resource name should not be affected after editing the duplicate's name.");
+	CHECK_MESSAGE(
+			Ref<Resource>(resource_dupe->get_meta("other_resource"))->get_name() == "My name was changed too",
+			"Duplicated resource should share its child resource with the original.");
+}
+
+TEST_CASE("[Resource] Saving and loading") {
+	Ref<Resource> resource = memnew(Resource);
+	resource->set_name("Hello world");
+	resource->set_meta("    ExampleMetadata    ", Vector2i(40, 80));
+	resource->set_meta("string", "The\nstring\nwith\nunnecessary\nline\n\t\\\nbreaks");
+	Ref<Resource> child_resource = memnew(Resource);
+	child_resource->set_name("I'm a child resource");
+	resource->set_meta("other_resource", child_resource);
+	const String save_path_binary = OS::get_singleton()->get_cache_path().plus_file("resource.res");
+	const String save_path_text = OS::get_singleton()->get_cache_path().plus_file("resource.tres");
+	ResourceSaver::save(save_path_binary, resource);
+	ResourceSaver::save(save_path_text, resource);
+
+	const Ref<Resource> loaded_resource_binary = ResourceLoader::load(save_path_binary);
+	CHECK_MESSAGE(
+			loaded_resource_binary->get_name() == "Hello world",
+			"The loaded resource name should be equal to the expected value.");
+	CHECK_MESSAGE(
+			loaded_resource_binary->get_meta("    ExampleMetadata    ") == Vector2i(40, 80),
+			"The loaded resource metadata should be equal to the expected value.");
+	CHECK_MESSAGE(
+			loaded_resource_binary->get_meta("string") == "The\nstring\nwith\nunnecessary\nline\n\t\\\nbreaks",
+			"The loaded resource metadata should be equal to the expected value.");
+	const Ref<Resource> loaded_child_resource_binary = loaded_resource_binary->get_meta("other_resource");
```

Comment:
I've seen `const Ref<T> &` is used (for bindings), not sure whether it would matter, but may be good for consistency:

```suggestion
	const Ref<Resource> &loaded_child_resource_binary = loaded_resource_binary->get_meta("other_resource");
```

The same applies to other `const Ref<Resource>` occurrences.